rr_common/src/planner/hill_climb_planner.cpp

```diff
@@ -0,0 +1,148 @@
+    if (plan.dists.size() != plan.path.size()) {
+        plan.dists.resize(plan.path.size());
+    }
+
+    plan.cost = 0;
+    plan.has_collision = false;
+    for (size_t i = 0; i < plan.path.size(); ++i) {
+        auto opt_dist = distance_checker_.GetCollisionDistance(plan.path[i].pose);
+        if (opt_dist) {
+            plan.cost += k_dist_ * std::exp(-std::max(0.01, opt_dist.value()));
+            plan.cost -= k_speed_ * plan.path[i].speed;
+            plan.cost += k_steering_ * plan.path[i].steer;
+            plan.cost += std::abs(plan.path[i].pose.theta) * k_angle_;
+            plan.dists[i] = opt_dist.value();
+        } else {
+            plan.cost += collision_penalty_ * (plan.path.size() - i);
+            std::fill(plan.dists.begin() + i, plan.dists.end(), 0);
+            plan.has_collision = true;
+            break;
+        }
+    }
+}
+
+void HillClimbPlanner::JitterControls(std::vector<double>& ctrl, double stddev) {
+    for (double& x : ctrl) {
+        x += normal_distribution_(rand_gen_) * stddev;
+        if (x < -max_steering_) {
+            x = -max_steering_;
+        } else if (x > max_steering_) {
+            x = max_steering_;
+        }
+    }
+}
+
+PlannedPath HillClimbPlanner::Plan(const rr::PCLMap& map) {
+    distance_checker_.SetMap(map);
+
+    auto descend_hill = [this, &map](PlannedPath& plan) {
+        plan.has_collision = false;
+        // if control is not inited externally, select a random starting configuration
+        if (plan.control.size() != segment_sections_.size()) {
+            plan.control.resize(segment_sections_.size());
+            std::fill(plan.control.begin(), plan.control.end(), 0.0);
+            JitterControls(plan.control, max_steering_ / 2);
+        }
+        double best_cost = std::numeric_limits<double>::max();
+
+        int stuck_counter = local_optimum_tries_;
+        int jitter_counter = 0;
+        while (stuck_counter > 0) {
+            const auto last_ctrl = plan.control;  // copy
+            JitterControls(plan.control, neighbor_stddev_);
+            jitter_counter++;
+            model_.RollOutPath(plan.control, plan.path);
+            FillObstacleCosts(plan);
+
+            if (plan.cost >= best_cost) {
+                --stuck_counter;
+                plan.control = last_ctrl;
+            } else {
+                stuck_counter = local_optimum_tries_;
+                best_cost = plan.cost;
+            }
+        }
+
+        // ROS_INFO_STREAM("descent took " << jitter_counter << " rollouts");
+    };
+
+    std::vector<PlannedPath> best_plans(num_workers_);
+    for (PlannedPath& plan : best_plans) {
+        plan.cost = std::numeric_limits<double>::max();
+    }
+    int plan_count = 0;
+    std::mutex plan_count_mutex;
+
+    auto worker = [&, this](int thread_idx) {
+        while (true) {
+            {
+                std::lock_guard lock(plan_count_mutex);
+                if (plan_count >= num_restarts_) {
```

Comment:
You could do this comparison before/without getting the lock since it is only a read

Reply:
I think there's a potential race condition if you check before the lock